File: allinone-kotlin/app/src/main/java/com/example/allinone/firebase/FirebaseIdManager.kt
package com.example.allinone.firebase

import android.util.Log
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await

/**
 * Manages sequential ID generation for Firebase Firestore documents.
 * This class handles retrieving and incrementing ID counters stored in the "counters" collection.
 */
class FirebaseIdManager {
    private val firestore = FirebaseFirestore.getInstance()
    private val countersCollection = firestore.collection("counters")

    /**
     * Gets the next sequential ID for a specific collection.
     * IDs start from 0 and increment sequentially.
     *
     * @param collectionName The name of the collection to get the next ID for
     * @return The next available ID as Long
     */
    suspend fun getNextId(collectionName: String): Long {
        try {
            // Use a standardized collection name for workout-related items
            val standardizedCollectionName = when (collectionName) {
                "programs" -> "workouts_programs"
                "workouts" -> "workouts_sessions"
                "exercises" -> "workouts_exercises"
                else -> collectionName
            }
            
            // Check if document exists
            val docRef = countersCollection.document(standardizedCollectionName)
            val doc = docRef.get().await()

            // If document doesn't exist, create it with initial count of 1
            if (!doc.exists() || doc.getLong("count") == null) {
                val initialData = mapOf("count" to 1L)
                docRef.set(initialData).await()
                return 1L
            }

            // Get the current count and atomically increment it
            docRef.update("count", FieldValue.increment(1)).await()
            
            // Retrieve the updated count
            val updatedDoc = docRef.get().await()
            val newCount = updatedDoc.getLong("count") ?: 1L
            
            return newCount
        } catch (e: Exception) {
            // Log failure and handle gracefully
            Log.e("FirebaseIdManager", "Failed to get next ID: ${e.message}")
            
            // In case of failure, generate a timestamp-based ID as fallback
            return System.currentTimeMillis()
        }
    }

    /**
     * Gets the last used ID for a specific collection without incrementing it.
     * This is useful for retrieving the current counter value without changing it.
     *
     * @param collectionName The name of the collection to get the last ID for
     * @return The last used ID as Long
     */
    suspend fun getLastId(collectionName: String): Long {
        try {
            // Get the current counter value without incrementing
            val counterDoc = countersCollection.document(collectionName).get().await()
            val count = counterDoc.getLong("count")

            return count ?: 0L
        } catch (e: Exception) {
            // If the counter doesn't exist, create it starting at 0
            if (e.message?.contains("NOT_FOUND") == true) {
                try {
                    // Create the counter document with initial value 0
                    val initialData = hashMapOf("count" to 0L)
                    countersCollection.document(collectionName).set(initialData).await()
                    return 0L // Return 0 as the first ID
                } catch (e2: Exception) {
                    Log.e(TAG, "Error creating counter for $collectionName", e2)
                    return 0L
                }
            } else {
                Log.e(TAG, "Error getting last ID for $collectionName", e)
                return 0L
            }
        }
    }

    companion object {
        private const val TAG = "FirebaseIdManager"
    }
}

File: src/data/firebaseIdManager.ts
import { doc, getDoc, setDoc, updateDoc, increment } from 'firebase/firestore';
import { getDb } from './firebase';

/**
 * Manages sequential ID generation for Firebase Firestore documents.
 * This class handles retrieving and incrementing ID counters stored in the "counters" collection.
 */
export class FirebaseIdManager {
  private db = getDb();
  private countersCollection = 'counters';

  /**
   * Gets the next sequential ID for a specific collection.
   * IDs start from 0 and increment sequentially.
   *
   * @param collectionName The name of the collection to get the next ID for
   * @returns The next available ID as number
   */
  async getNextId(collectionName: string): Promise<number> {
    try {
      // Use a standardized collection name for workout-related items
      const standardizedCollectionName = this.getStandardizedCollectionName(collectionName);
      console.log('FirebaseIdManager: Getting next ID for', collectionName, '->', standardizedCollectionName);
      
      // Check if document exists
      const docRef = doc(this.db, this.countersCollection, standardizedCollectionName);
      const docSnap = await getDoc(docRef);

      // If document doesn't exist, create it with initial count of 1
      if (!docSnap.exists() || !docSnap.data()?.count) {
        const initialData = { count: 1 };
        await setDoc(docRef, initialData);
        console.log('FirebaseIdManager: Created new counter for', standardizedCollectionName, 'with ID 1');
        return 1;
      }

      // Get the current count and atomically increment it
      await updateDoc(docRef, {
        count: increment(1)
      });
      
      // Retrieve the updated count
      const updatedDocSnap = await getDoc(docRef);
      const newCount = updatedDocSnap.data()?.count || 1;
      
      console.log('FirebaseIdManager: Generated ID', newCount, 'for', standardizedCollectionName);
      return newCount;
    } catch (error) {
      // Log failure and handle gracefully
      console.error('FirebaseIdManager: Failed to get next ID:', error);
      
      // In case of failure, generate a timestamp-based ID as fallback
      const fallbackId = Date.now();
      console.log('FirebaseIdManager: Using fallback ID:', fallbackId);
      return fallbackId;
    }
  }

  /**
   * Gets the last used ID for a specific collection without incrementing it.
   * This is useful for retrieving the current counter value without changing it.
   *
   * @param collectionName The name of the collection to get the last ID for
   * @returns The last used ID as number
   */
  async getLastId(collectionName: string): Promise<number> {
    try {
      // Get the current counter value without incrementing
      const docRef = doc(this.db, this.countersCollection, collectionName);
      const docSnap = await getDoc(docRef);
      const count = docSnap.data()?.count;

      return count || 0;
    } catch (error) {
      // If the counter doesn't exist, create it starting at 0
      if (error instanceof Error && error.message.includes('NOT_FOUND')) {
        try {
          // Create the counter document with initial value 0
          const initialData = { count: 0 };
          const docRef = doc(this.db, this.countersCollection, collectionName);
          await setDoc(docRef, initialData);
          return 0; // Return 0 as the first ID
        } catch (createError) {
          console.error('FirebaseIdManager: Error creating counter for', collectionName, createError);
          return 0;
        }
      } else {
        console.error('FirebaseIdManager: Error getting last ID for', collectionName, error);
        return 0;
      }
    }
  }

  /**
   * Standardizes collection names for workout-related items to match the Kotlin implementation.
   */
  private getStandardizedCollectionName(collectionName: string): string {
    switch (collectionName) {
      case 'programs':
        return 'workouts_programs';
      case 'workouts':
        return 'workouts_sessions';
      case 'exercises':
        return 'workouts_exercises';
      case 'tasks':
        return 'tasks';
      case 'taskGroups':
        return 'taskGroups';
      default:
        return collectionName;
    }
  }
}

// Export a singleton instance
export const firebaseIdManager = new FirebaseIdManager(); 